design-proposals/klip-10-suppress.md

```diff
@@ -1 +1,123 @@
+           refinement policy of underlying KStreams applications. `KTable#suppress` allows such
+           control and should be given corresponding KSQL syntax._
+           
+## Motivation and background
+
+Inspired by: [Kafka Streams' Take on Watermarks and Triggers](
+https://www.confluent.io/blog/kafka-streams-take-on-watermarks-and-triggers)
+
+KSQL provides mechanisms to aggregate data and maintain an internal materialization. As a motivating
+example, let's consider an alerting mechanism built on KSQL. We begin with a stream `healthcheck` 
+that has the schema `server VARCHAR`. Servers emit to this stream every minute with their own name
+to indicate that they are up and healthy. We create a KSQL application from this stream:
+
+```sql 
+CREATE TABLE alerts AS
+    SELECT COUNT(*) as count, server
+    FROM healthcheck
+    WINDOW TUMBLING (SIZE 5 MINUTES) 
+    GROUP BY server
+    EMIT CHANGES;
+```
+
+We have a downstream microservice that consumes from the `alerts` topic and sends an email whenever
+the count is less than 3 for a server (i.e. a server missed 2 out of 5 heartbeat healthchecks in a
+window of 5 minutes). 
+
+Today, building such an application would not work the way you want it to! The downstream service
+would get intermediate results (the exact results are slightly unpredictable, as they depend on
+the `commit.interval.ms` and `cache.max.bytes.buffering` configurations), and would spuriously
+alert. For example, if the `healtcheck` stream had the following data:
+```
+ts: min_1 server: abc
+ts: min_2 server: abc
+ts: min_3 server: abc
+```
+the `alerts` topic would get the update three times (each time with incrementing `count` column 
+value), and would send three emails when none should be sent.
+
+## Scope
+
+**In Scope**: 
+- Outline the features of suppress that need to be represented in KSQL
+- Syntax for suppress functionality
+- Outline of potential implementation
+
+**Out of Scope**:
+- Grace periods for windows
+- Custom suppressions
+
+## Value/Return
+
+This feature enables functionality in KSQL that is not possible today, and provides stricter
+semantic guarantees on KSQL aggregation queries.
+
+## Public APIS
+
+The public API will extend the `EMIT` clause to support refinement operators other than `CHANGES`
+(which is the only currently supported refinement). The proposed syntax is:
+
+```sql
+SELECT select_expression
+    FROM source
+    WINDOW window_expression
+    GROUP BY grouping_expression
+    EMIT refinement_expression
+```
+
+A `refinement_expression` has the following syntax:
+```sql
+CHANGES | FINAL [EAGERLY] [WHEN suppression]
+```
+
+And behaves in the following ways:
+- If `CHANGES` is specified, then all intermediate changes will be materialized
+- If `FINAL` is specified, then output will be suppressed depending on the `suppression`:
+    - if `EAGERLY` is specified, the suppression is a best effort attempt to reduce duplicate output
```

Comment:
What is the difference between `CHANGES ` and `FINAL EAGERLY`? To me, they appear to be the same.

Reply:
Ah I was under the assumption that `CHANGES` would be "one-to-one" (e.g. something like no buffering) - I've had some people in the community ask for that feature. Is that possible with the suppress API?

Reply:
After discussing with @vpapavas a little more - I think I've clarified my own thoughts about the difference:
- `EMIT CHANGES` uses just the buffer size
- `EMIT FINAL EAGERLY` would use either the buffer size or the window size, whatever comes first

Reply:
are we set on `EMIT FINAL` ?  I goes back to https://github.com/confluentinc/ksql/pull/3242#issuecomment-528120096  

I am concerned about introducing a new `FINAL` mode here. Could we do something like adding a new  `SUPRESS UNTIL <supression_mechanism>` clause ? (`SUPRESS 
- IMO having supress in the syntax would also make it easier for users to reason across streams and ksql. Its anyway doing a Streams supress
- Can EAGERLY be a supression scheme instead of its own clause.. For with `EAGERLY WHEN WINDOW_CLOSED()` , user would be 1 record per window as long as it fits in buffer?

Reply:
Look at the comments below that John posted - I think it helps clarify a lot of the streams behaviors that will make me need to revamp a lot of this. 

> Are we set on `EMIT FINAL`?

I think we can change `FINAL` to something else if we want, but I think we should stick with `EMIT` and not introduce `SUPPRESS` - I don't agree that mirroring Streams terminology is necessarily a design goal. One candidate is maybe `EMIT WHEN [supression_condition]`.

> Can EAGERLY be a suppression scheme instead of its own clause

I think I'm going to remove the concept of `EAGERLY` - it looks like I misunderstood how that works. Again, I will defer to John's comments below on the semantics that I misunderstood.

Reply:
`EMIT FINAL` also fits in with the direction other streaming SQL impls are going... i.e. Beam/Flink.